lnbits/extensions/scheduledpayments/tasks.py

```diff
@@ -0,0 +1,185 @@
+
+    for schedule in schedules:
+        # run an individual schedule
+        result = await run_scheduled_payment(schedule)
+
+    return True
+
+
+async def run_scheduled_payment(schedule):
+    now = pendulum.now(schedule.timezone)
+
+    # determine if the schedule starts in the future. If it does, skip
+    if schedule.start_date:
+        start_date = pendulum.parse(schedule.start_date, tz=schedule.timezone)
+
+        if start_date > now:
+            return True
+
+    # determine if the schedule ends in the past. If it does, skip
+    if schedule.end_date:
+        end_date = pendulum.parse(schedule.end_date, tz=schedule.timezone)
+
+        if end_date < now:
+            return True
+
+    # fetch all events for this schedule
+    events = await get_schedule_events(schedule.id)
+
+    # if no previous events, go back to 0
+    if not events:
+        lastrun = pendulum.from_timestamp(0, schedule.timezone)
+    else:
+        # sort the list by time desc
+        sortedevents = sorted(
+            events,
+            key=lambda d: d.time,
+            reverse=True,
+        )
+
+        lastevent = sortedevents[0]
+
+        # the most recent event is now first in the list
+        lastrun = pendulum.from_timestamp(lastevent.time, schedule.timezone)
+
+        # if the last event is still in_progress, we should investigate
+        if lastevent.status == "in_progress":
+            # if the last run started > 5 minutes ago we should update the status
+            if now > lastrun.add(minutes=5):
+                lastevent.status = "Error: Timed Out"
+                event = await update_schedule_event(data=lastevent)
+
+            return True
+
+    cron = ct.Crontab.parse(
+        schedule.interval,
+        tz=timezone(schedule.timezone),
+    )
+
+    # calculate when the next run should be
+    nextrun = cron.next_fire_time(now=lastrun)
+
+    # determine whether we should execute this schedule now
+    if nextrun < now:
+        print(f"{nextrun} < {now} fire!")
+
+        amount_in_sats = await fiat_amount_as_satoshis(
+            schedule.famount / 100, schedule.currency
+        )
+
+        event = await create_schedule_event(
+            data={
+                "schedule_id": schedule.id,
+                "amount": amount_in_sats,
+                "status": "in_progress",
+            }
+        )
+
+        # users can enter bech32 encoded lnurl, or a lightning address like lnbits@lntips.com
+        if schedule.recipient.upper().startswith("LNURL"):
+            url = decode_lnurl(schedule.recipient)
+        else:
+            parts = schedule.recipient.split("@")
+
+            if len(parts) != 2:
+                event.status = f"Error: Bad Lightning Address"
+                event = await update_schedule_event(data=event)
+                return False
+
+            username = parts[0]
+            domain = parts[1]
+            url = f"https://{domain}/.well-known/lnurlp/{username}"
+
```

Comment:
i think somewhere here it would be useful, to check the balance of the wallet upfront, before we requesting an invoice from lnurl.

Reply:
@dni I've just pushed up a balance check here.